File: ngs/ngs-java/ngs/itf/AlignmentItf.java
/*===========================================================================
*
*                            PUBLIC DOMAIN NOTICE
*               National Center for Biotechnology Information
*
*  This software/database is a "United States Government Work" under the
*  terms of the United States Copyright Act.  It was written as part of
*  the author's official duties as a United States Government employee and
*  thus cannot be copyrighted.  This software/database is freely available
*  to the public for use. The National Library of Medicine and the U.S.
*  Government have not placed any restriction on its use or reproduction.
*
*  Although all reasonable efforts have been taken to ensure the accuracy
*  and reliability of the software and data, the NLM and the U.S.
*  Government do not and cannot warrant the performance or results that
*  may be obtained by using this software or data. The NLM and the U.S.
*  Government disclaim all warranties, express or implied, including
*  warranties of performance, merchantability or fitness for any particular
*  purpose.
*
*  Please cite the author in any work or product based on this material.
*
* ===========================================================================
*
*/

package ngs.itf;

import ngs.ErrorMsg;
import ngs.Fragment;
import ngs.Alignment;


/*==========================================================================
 * AlignmentItf
 *  represents an alignment between a Fragment and Reference sub-sequence
 *  provides a path to Read and mate Alignment
 */
class AlignmentItf
    extends Refcount
    implements Alignment
{

    /**********************
     * Fragment Interface *
     **********************/

    /* getFragmentId
     */
    public String getFragmentId ()
        throws ErrorMsg
    {
        return this . GetFragmentId ( self );
    }

    /* getFragmentBases
     *  return sequence bases
     *  "offset" is zero-based
     *  "length" must be >= 0
     */
    public String getFragmentBases ()
        throws ErrorMsg
    {
        return this . GetFragmentBases ( self, 0, -1 );
    }

    public String getFragmentBases ( long offset )
        throws ErrorMsg, IndexOutOfBoundsException
    {
        if ( offset < 0 )
            throw new IndexOutOfBoundsException ( "offset " + offset + " is negative" );

        return this . GetFragmentBases ( self, offset, -1 );
    }

    public String getFragmentBases ( long offset, long length )
        throws ErrorMsg, IndexOutOfBoundsException
    {
        if ( offset < 0 )
            throw new IndexOutOfBoundsException ( "offset " + offset + " is negative" );
        if ( length < 0 )
            throw new IndexOutOfBoundsException ( "length " + length + " is negative");

        return this . GetFragmentBases ( self, offset, length );
    }


    /* getFragmentQualities
     *  return phred quality values
     *  using ASCII offset of 33
     *  "offset" is zero-based
     *  "length" must be >= 0
     */
    public String getFragmentQualities ()
        throws ErrorMsg
    {
        return this . GetFragmentQualities ( self, 0, -1 );
    }

    public String getFragmentQualities ( long offset )
        throws ErrorMsg, IndexOutOfBoundsException
    {
        if ( offset < 0 )
            throw new IndexOutOfBoundsException ( "offset " + offset + " is negative" );

        return this . getFragmentQualities ( offset, -1 );
    }

    public String getFragmentQualities ( long offset, long length )
        throws ErrorMsg, IndexOutOfBoundsException
    {
        if ( offset < 0 )
            throw new IndexOutOfBoundsException ( "offset " + offset + " is negative" );
        if ( length < 0 )
            throw new IndexOutOfBoundsException ( "length " + length + " is negative");

        return this . GetFragmentQualities ( self, offset, length );
    }

    public boolean isPaired ()
        throws ErrorMsg
    {
        return this . IsPaired ( self );
    }
    
    public boolean isAligned ()
        throws ErrorMsg
    {
        return true;
    }

    /***********************
     * Alignment Interface *
     ***********************/

    /* getAlignmentId
     *  unique within ReadCollection
     */
    public String getAlignmentId ()
        throws ErrorMsg
    {
        return this . GetAlignmentId ( self );
    }


    /*------------------------------------------------------------------
     * Reference
     */

    /* getReferenceSpec
     */
    public String getReferenceSpec ()
        throws ErrorMsg
    {
        return this . GetReferenceSpec ( self );
    }

    /* getMappingQuality 
     */
    public int getMappingQuality ()
        throws ErrorMsg
    {
        return this . GetMappingQuality ( self );
    }

    /* getReferenceBases
     *  return reference bases
     */
    public String getReferenceBases ()
        throws ErrorMsg
    {
        return this . GetReferenceBases ( self );
    }


    /*------------------------------------------------------------------
     * Fragment
     */

    /* getReadGroup
     */
    public String getReadGroup ()
        throws ErrorMsg
    {
        return this . GetReadGroup ( self );
    }

    /* getReadId
     */
    public String getReadId ()
        throws ErrorMsg
    {
        return this . GetReadId ( self );
    }

    /* getClippedFragmentBases
     *  return fragment bases
     */
    public String getClippedFragmentBases ()
        throws ErrorMsg
    {
        return this . GetClippedFragmentBases ( self );
    }

    /* getClippedFragmentQualities
     *  return fragment phred quality values
     *  using ASCII offset of 33
     */
    public String getClippedFragmentQualities ()
        throws ErrorMsg
    {
        return this . GetClippedFragmentQualities ( self );
    }

    /* getAlignedFragmentBases
     *  return fragment bases in their aligned orientation
     */
    public String getAlignedFragmentBases ()
        throws ErrorMsg
    {
        return this . GetAlignedFragmentBases ( self );
    }
        
    /*------------------------------------------------------------------
     * details of this alignment
     */


    /* getAlignmentCategory
     */
    public int getAlignmentCategory ()
        throws ErrorMsg
    {
        return this . GetAlignmentCategory ( self );
    }

    /* getAlignmentPosition
     *  returns the alignment's starting position on reference
     */
    public long getAlignmentPosition ()
        throws ErrorMsg
    {
        return this . GetAlignmentPosition ( self );
    }

    /* getAligmentLength
     *  returns the alignment's projected length upon reference
     */
    public long getAlignmentLength ()
        throws ErrorMsg
    {
        return this . GetAlignmentLength ( self );
    }

    /* getIsReversedOrientation
     *  returns true if orientation is reversed
     *  with respect to the reference sequence
     */
    public boolean getIsReversedOrientation ()
        throws ErrorMsg
    {
        return this . GetIsReversedOrientation ( self );
    }

    /* getSoftClip 
     */
    public int getSoftClip ( int edge )
        throws ErrorMsg
    {
        return this . GetSoftClip ( self, edge );
    }

    /* getTemplateLength
     */
    public long getTemplateLength ()
        throws ErrorMsg
    {
        return this . GetTemplateLength ( self );
    }

    /* getShortCigar
     *  returns a text string describing alignment details
     */
    public String getShortCigar ( boolean clipped )
        throws ErrorMsg
    {
        return this . GetShortCigar ( self, clipped );
    }

    /* getLongCigar
     *  returns a text string describing alignment details
     */
    public String getLongCigar ( boolean clipped )
        throws ErrorMsg
    {
        return this . GetLongCigar ( self, clipped );
    }


    /* getRNAOrientation
     */
    public char getRNAOrientation ()
        throws ErrorMsg
    {
        return this . GetRNAOrientation ( self );
    }



    /*------------------------------------------------------------------
     * details of mate alignment
     */

    /* hasMate
     */
    public boolean hasMate ()
    {
        return this . HasMate ( self );
    }
    
    /* getMateAlignmentId
     */
    public String getMateAlignmentId ()
        throws ErrorMsg
    {
        return this . GetMateAlignmentId ( self );
    }

    /* getMateAlignment
     */
    public Alignment getMateAlignment ()
        throws ErrorMsg
    {
        long ref = this . GetMateAlignment ( self );
        try
        {
            return new AlignmentItf ( ref );
        }
        catch ( Exception x )
        {
            this . release ( ref );
            throw new ErrorMsg ( x . toString () );
        }
    }

    /* getMateReferenceSpec
     */
    public String getMateReferenceSpec ()
        throws ErrorMsg
    {
        return this . GetMateReferenceSpec ( self );
    }

    /* getMateIsReversedOrientation
     */
    public boolean getMateIsReversedOrientation ()
        throws ErrorMsg
    {
        return this . GetMateIsReversedOrientation ( self );
    }


    /*******************************
     * AlignmentItf Implementation *
     *******************************/


    // constructors
    AlignmentItf ( long ref )
    {
        super ( ref );
    }

    AlignmentItf ( Alignment obj )
        throws ErrorMsg
    {
        super ( 0 );
        try
        {
            AlignmentItf ref = ( AlignmentItf ) obj;
            this . self = ref . duplicate ();
        }
        catch ( Exception x )
        {
            throw new ErrorMsg ( x . toString () );
        }
    }

    // native interface
    private native String GetFragmentId ( long self )
        throws ErrorMsg;
    private native String GetFragmentBases ( long self, long offset, long length )
        throws ErrorMsg;
    private native String GetFragmentQualities ( long self, long offset, long length )
        throws ErrorMsg;
    private native boolean IsPaired ( long self )
        throws ErrorMsg;
    private native String GetAlignmentId ( long self )
        throws ErrorMsg;
    private native String GetReferenceSpec ( long self )
        throws ErrorMsg;
    private native int GetMappingQuality ( long self )
        throws ErrorMsg;
    private native String GetReferenceBases ( long self )
        throws ErrorMsg;
    private native String GetReadGroup ( long self )
        throws ErrorMsg;
    private native String GetReadId ( long self )
        throws ErrorMsg;
    private native String GetClippedFragmentBases ( long self )
        throws ErrorMsg;
    private native String GetClippedFragmentQualities ( long self )
        throws ErrorMsg;
    private native String GetAlignedFragmentBases ( long self )
        throws ErrorMsg;
    private native int GetAlignmentCategory ( long self )
        throws ErrorMsg;
    private native long GetAlignmentPosition ( long self )
        throws ErrorMsg;
    private native long GetAlignmentLength ( long self )
        throws ErrorMsg;
    private native boolean GetIsReversedOrientation ( long self )
        throws ErrorMsg;
    private native int GetSoftClip ( long self, int edge )
        throws ErrorMsg;
    private native long GetTemplateLength ( long self )
        throws ErrorMsg;
    private native String GetShortCigar ( long self, boolean clipped )
        throws ErrorMsg;
    private native String GetLongCigar ( long self, boolean clipped )
        throws ErrorMsg;
    private native char GetRNAOrientation ( long self )
        throws ErrorMsg;
    private native boolean HasMate ( long self );
    private native String GetMateAlignmentId ( long self )
        throws ErrorMsg;
    private native long GetMateAlignment ( long self )
        throws ErrorMsg;
    private native String GetMateReferenceSpec ( long self )
        throws ErrorMsg;
    private native boolean GetMateIsReversedOrientation ( long self )
        throws ErrorMsg;
}

File: ngs/ngs-sdk/adapter/AlignmentItf.cpp
/*===========================================================================
*
*                            PUBLIC DOMAIN NOTICE
*               National Center for Biotechnology Information
*
*  This software/database is a "United States Government Work" under the
*  terms of the United States Copyright Act.  It was written as part of
*  the author's official duties as a United States Government employee and
*  thus cannot be copyrighted.  This software/database is freely available
*  to the public for use. The National Library of Medicine and the U.S.
*  Government have not placed any restriction on its use or reproduction.
*
*  Although all reasonable efforts have been taken to ensure the accuracy
*  and reliability of the software and data, the NLM and the U.S.
*  Government do not and cannot warrant the performance or results that
*  may be obtained by using this software or data. The NLM and the U.S.
*  Government disclaim all warranties, express or implied, including
*  warranties of performance, merchantability or fitness for any particular
*  purpose.
*
*  Please cite the author in any work or product based on this material.
*
* ===========================================================================
*
*/

#include <ngs/adapter/AlignmentItf.hpp>
#include <ngs/adapter/StringItf.hpp>
#include <ngs/adapter/ErrorMsg.hpp>

#include "ErrBlock.hpp"

namespace ngs_adapt
{

    /*----------------------------------------------------------------------
     * AlignmentItf
     */

    AlignmentItf :: AlignmentItf ()
        : FragmentItf ( & ivt . dad )
    {
    }

    bool AlignmentItf :: nextFragment ()
    {
        throw ErrorMsg ( "INTERNAL ERROR: nextFragment message on Alignment" );
    }

    NGS_String_v1 * CC AlignmentItf :: get_id ( const NGS_Alignment_v1 * iself, NGS_ErrBlock_v1 * err )
    {
        const AlignmentItf * self = Self ( iself );
        try
        {
            StringItf * val = self -> getAlignmentId ();
            return val -> Cast ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    NGS_String_v1 * CC AlignmentItf :: get_ref_spec ( const NGS_Alignment_v1 * iself, NGS_ErrBlock_v1 * err )
    {
        const AlignmentItf * self = Self ( iself );
        try
        {
            StringItf * val = self -> getReferenceSpec ();
            return val -> Cast ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    int32_t CC AlignmentItf :: get_map_qual ( const NGS_Alignment_v1 * iself, NGS_ErrBlock_v1 * err )
    {
        const AlignmentItf * self = Self ( iself );
        try
        {
            return self -> getMappingQuality ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    NGS_String_v1 * CC AlignmentItf :: get_ref_bases ( const NGS_Alignment_v1 * iself, NGS_ErrBlock_v1 * err )
    {
        const AlignmentItf * self = Self ( iself );
        try
        {
            StringItf * val = self -> getReferenceBases ();
            return val -> Cast ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    NGS_String_v1 * CC AlignmentItf :: get_read_group ( const NGS_Alignment_v1 * iself, NGS_ErrBlock_v1 * err )
    {
        const AlignmentItf * self = Self ( iself );
        try
        {
            StringItf * val = self -> getReadGroup ();
            return val -> Cast ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    NGS_String_v1 * CC AlignmentItf :: get_read_id ( const NGS_Alignment_v1 * iself, NGS_ErrBlock_v1 * err )
    {
        const AlignmentItf * self = Self ( iself );
        try
        {
            StringItf * val = self -> getReadId ();
            return val -> Cast ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    NGS_String_v1 * CC AlignmentItf :: get_clipped_frag_bases ( const NGS_Alignment_v1 * iself, NGS_ErrBlock_v1 * err )
    {
        const AlignmentItf * self = Self ( iself );
        try
        {
            StringItf * val = self -> getClippedFragmentBases ();
            return val -> Cast ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    NGS_String_v1 * CC AlignmentItf :: get_clipped_frag_quals ( const NGS_Alignment_v1 * iself, NGS_ErrBlock_v1 * err )
    {
        const AlignmentItf * self = Self ( iself );
        try
        {
            StringItf * val = self -> getClippedFragmentQualities ();
            return val -> Cast ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    NGS_String_v1 * CC AlignmentItf :: get_aligned_frag_bases ( const NGS_Alignment_v1 * iself, NGS_ErrBlock_v1 * err )
    {
        const AlignmentItf * self = Self ( iself );
        try
        {
            StringItf * val = self -> getAlignedFragmentBases ();
            return val -> Cast ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    bool CC AlignmentItf :: is_primary ( const NGS_Alignment_v1 * iself, NGS_ErrBlock_v1 * err )
    {
        const AlignmentItf * self = Self ( iself );
        try
        {
            return self -> isPrimary ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    int64_t CC AlignmentItf :: get_align_pos ( const NGS_Alignment_v1 * iself, NGS_ErrBlock_v1 * err )
    {
        const AlignmentItf * self = Self ( iself );
        try
        {
            return self -> getAlignmentPosition ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    uint64_t CC AlignmentItf :: get_ref_pos_projection_range ( const NGS_Alignment_v1 * iself, NGS_ErrBlock_v1 * err, int64_t ref_pos )
    {
        const AlignmentItf * self = Self ( iself );
        try
        {
            return self -> getReferencePositionProjectionRange ( ref_pos );
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    uint64_t CC AlignmentItf :: get_align_length ( const NGS_Alignment_v1 * iself, NGS_ErrBlock_v1 * err )
    {
        const AlignmentItf * self = Self ( iself );
        try
        {
            return self -> getAlignmentLength ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    bool CC AlignmentItf :: get_is_reversed ( const NGS_Alignment_v1 * iself, NGS_ErrBlock_v1 * err )
    {
        const AlignmentItf * self = Self ( iself );
        try
        {
            return self -> getIsReversedOrientation ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return false;
    }

    int32_t CC AlignmentItf :: get_soft_clip ( const NGS_Alignment_v1 * iself, NGS_ErrBlock_v1 * err, uint32_t edge )
    {
        const AlignmentItf * self = Self ( iself );
        try
        {
            return self -> getSoftClip ( edge );
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    uint64_t CC AlignmentItf :: get_template_len ( const NGS_Alignment_v1 * iself, NGS_ErrBlock_v1 * err )
    {
        const AlignmentItf * self = Self ( iself );
        try
        {
            return self -> getTemplateLength ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    NGS_String_v1 * CC AlignmentItf :: get_short_cigar ( const NGS_Alignment_v1 * iself, NGS_ErrBlock_v1 * err, bool clipped )
    {
        const AlignmentItf * self = Self ( iself );
        try
        {
            StringItf * val = self -> getShortCigar ( clipped );
            return val -> Cast ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    NGS_String_v1 * CC AlignmentItf :: get_long_cigar ( const NGS_Alignment_v1 * iself, NGS_ErrBlock_v1 * err, bool clipped )
    {
        const AlignmentItf * self = Self ( iself );
        try
        {
            StringItf * val = self -> getLongCigar ( clipped );
            return val -> Cast ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    char CC AlignmentItf :: get_rna_orientation ( const NGS_Alignment_v1 * iself, NGS_ErrBlock_v1 * err )
    {
        const AlignmentItf * self = Self ( iself );
        try
        {
            return self -> getRNAOrientation ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return false;
    }

    bool CC AlignmentItf :: has_mate ( const NGS_Alignment_v1 * iself, NGS_ErrBlock_v1 * err )
    {
        const AlignmentItf * self = Self ( iself );
        try
        {
            return self -> hasMate ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return false;
    }

    NGS_String_v1 * CC AlignmentItf :: get_mate_id ( const NGS_Alignment_v1 * iself, NGS_ErrBlock_v1 * err )
    {
        const AlignmentItf * self = Self ( iself );
        try
        {
            StringItf * val = self -> getMateAlignmentId ();
            return val -> Cast ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    NGS_Alignment_v1 * CC AlignmentItf :: get_mate_alignment ( const NGS_Alignment_v1 * iself, NGS_ErrBlock_v1 * err )
    {
        const AlignmentItf * self = Self ( iself );
        try
        {
            AlignmentItf * val = self -> getMateAlignment ();
            return val -> Cast ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    NGS_String_v1 * CC AlignmentItf :: get_mate_ref_spec ( const NGS_Alignment_v1 * iself, NGS_ErrBlock_v1 * err )
    {
        const AlignmentItf * self = Self ( iself );
        try
        {
            StringItf * val = self -> getMateReferenceSpec ();
            return val -> Cast ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    bool CC AlignmentItf :: get_mate_is_reversed ( const NGS_Alignment_v1 * iself, NGS_ErrBlock_v1 * err )
    {
        const AlignmentItf * self = Self ( iself );
        try
        {
            return self -> getMateIsReversedOrientation ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return false;
    }

    bool CC AlignmentItf :: next ( NGS_Alignment_v1 * iself, NGS_ErrBlock_v1 * err )
    {
        AlignmentItf * self = Self ( iself );
        try
        {
            return self -> nextAlignment ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return false;
    }

    NGS_Alignment_v1_vt AlignmentItf :: ivt =
    {
        {
            "ngs_adapt::AlignmentItf",
            "NGS_Alignment_v1",
            1,
            & FragmentItf :: ivt . dad
        },

        // v1.0
        get_id,
        get_ref_spec,
        get_map_qual,
        get_ref_bases,
        get_read_group,
        get_read_id,
        get_clipped_frag_bases,
        get_clipped_frag_quals,
        get_aligned_frag_bases,
        is_primary,
        get_align_pos,
        get_align_length,
        get_is_reversed,
        get_soft_clip,
        get_template_len,
        get_short_cigar,
        get_long_cigar,
        has_mate,
        get_mate_id,
        get_mate_alignment,
        get_mate_ref_spec,
        get_mate_is_reversed,
        next,

        // v1.1
        get_rna_orientation,

        // v1.2
        get_ref_pos_projection_range
    };

} // namespace ngs_adapt
